packages/core/src/pubsub/incoming-channel.ts

```diff
@@ -0,0 +1,107 @@
+import { interval, Observable } from 'rxjs';
+import { Memoize } from 'typescript-memoize';
+import { IpfsApi } from '@ceramicnetwork/common';
+import { TaskQueue } from './task-queue';
+import { DiagnosticsLogger, ServiceLogger } from '@ceramicnetwork/logger';
+import { pipe, MonoTypeOperatorFunction } from 'rxjs';
+import { map, filter, concatMap } from 'rxjs/operators';
+
+// Typestub for pubsub message.
+// At some future time this type definition should be provided by IPFS.
+export type IPFSPubsubMessage = {
+  from: string;
+  seqno: Uint8Array;
+  data: Uint8Array;
+  topicIDs: string[];
+  signature: Uint8Array;
+  key: Uint8Array;
+};
+
+async function resubscribe(
+  ipfs: IpfsApi,
+  topic: string,
+  pubsubLogger: ServiceLogger,
+  handler: (message: IPFSPubsubMessage) => void,
+) {
+  const listeningTopics = await ipfs.pubsub.ls();
+  const isSubscribed = listeningTopics.includes(topic);
+  if (!isSubscribed) {
+    await ipfs.pubsub.unsubscribe(topic, handler);
+    await ipfs.pubsub.subscribe(topic, handler);
+    const ipfsId = await ipfs.id();
+    const peerId = ipfsId.id;
+    pubsubLogger.log({ peer: peerId, event: 'subscribed', topic: topic });
+  }
+}
+
+/**
+ * Incoming IPFS PubSub message stream as Observable.
+ * Ensures that ipfs subscription to a topic is live by periodically re-subscribing.
+ *
+ * One could simplify this by removing periodic re-subscription and serialization via task-queue.
+ * Unfortunately, we can end up with concurrency issues: trying to unsubscribe while (or right after) other instance tries to subscribe.
+ * So, better keep it all together.
+ */
+export class IncomingChannel extends Observable<IPFSPubsubMessage> {
+  constructor(
+    ipfs: IpfsApi,
+    topic: string,
+    resubscribeEvery: number,
+    pubsubLogger: ServiceLogger,
+    readonly logger: DiagnosticsLogger,
+  ) {
+    super((subscriber) => {
+      const handler = (message: IPFSPubsubMessage) => subscriber.next(message);
+
+      this.tasks.add(() => resubscribe(ipfs, topic, pubsubLogger, handler));
+
+      const ensureSubscribed = interval(resubscribeEvery).subscribe(() => {
+        this.tasks.add(() => resubscribe(ipfs, topic, pubsubLogger, handler));
+      });
+
+      return () => {
+        ensureSubscribed.unsubscribe();
+        // Remove pending subscription attempts.
+        this.tasks.clear();
```

Comment:
Could more tasks get added in between this `clear()` and when the unsubscribe task added below actually runs?

Reply:
The only thing can do this is controlled by `ensureSubscribed`. After it is effectively stopped (`ensureSubscribed.unsubscribe`), nothing can interfere.

Reply:
Ah I see.  And this is because `tasks` isn't actually used for incoming pub/sub messages, only to serialize resubscribe attempts.  Okay makes sense